RecoEgamma/EgammaPhotonProducers/src/ReducedEGProducer.cc

```diff
@@ -847,55 +822,48 @@ void ReducedEGProducer::produce(edm::Event& theEvent, const edm::EventSetup& the
     relinkSuperCluster(gsfElectronCore, superClusterMap, outSuperClusterHandle);
     relinkGsfTrack(gsfElectronCore, gsfTrackMap, outGsfTrackHandle);
   }
 
   //put ootsuperclusters in the event
   edm::OrphanHandle<reco::SuperClusterCollection> outOOTSuperClusterHandle;
   if (!ootPhotonT_.isUninitialized())
-    outOOTSuperClusterHandle = theEvent.put(std::move(ootSuperClusters), outOOTSuperClusters_);
+    outOOTSuperClusterHandle = event.emplace(outOOTSuperClusters_, std::move(ootSuperClusters));
 
   //Relink OOTPhoton SuperClusters
-  for (reco::PhotonCore& ootPhotonCore : *ootPhotonCores) {
+  for (reco::PhotonCore& ootPhotonCore : ootPhotonCores) {
     relinkSuperCluster(ootPhotonCore, ootSuperClusterMap, outOOTSuperClusterHandle);
   }
 
   //put photoncores and gsfelectroncores into the event
-  const auto& outPhotonCoreHandle = theEvent.put(std::move(photonCores), outPhotonCores_);
+  const auto& outPhotonCoreHandle = event.emplace(outPhotonCores_, std::move(photonCores));
   edm::OrphanHandle<reco::PhotonCoreCollection> outOOTPhotonCoreHandle;
   if (!ootPhotonT_.isUninitialized())
-    outOOTPhotonCoreHandle = theEvent.put(std::move(ootPhotonCores), outOOTPhotonCores_);
-  const auto& outgsfElectronCoreHandle = theEvent.put(std::move(gsfElectronCores), outGsfElectronCores_);
+    outOOTPhotonCoreHandle = event.emplace(outOOTPhotonCores_, std::move(ootPhotonCores));
+  const auto& outgsfElectronCoreHandle = event.emplace(outGsfElectronCores_, std::move(gsfElectronCores));
 
   //loop over photons, oot photons, and electrons and relink the cores
-  for (reco::Photon& photon : *photons) {
+  for (reco::Photon& photon : photons) {
     relinkPhotonCore(photon, photonCoreMap, outPhotonCoreHandle);
   }
 
   if (!ootPhotonT_.isUninitialized()) {
-    for (reco::Photon& ootPhoton : *ootPhotons) {
+    for (reco::Photon& ootPhoton : ootPhotons) {
       relinkPhotonCore(ootPhoton, ootPhotonCoreMap, outOOTPhotonCoreHandle);
     }
   }
 
-  for (reco::GsfElectron& gsfElectron : *gsfElectrons) {
+  for (reco::GsfElectron& gsfElectron : gsfElectrons) {
     relinkGsfElectronCore(gsfElectron, gsfElectronCoreMap, outgsfElectronCoreHandle);
 
-    // -----
-    // Also in this loop let's relink ambiguous tracks
-    std::vector<reco::GsfTrackRef> ambigTracksInThisElectron;
-    // Here we loop over the ambiguous tracks and save them in a vector
-    for (auto igsf = gsfElectron.ambiguousGsfTracksBegin(); igsf != gsfElectron.ambiguousGsfTracksEnd(); ++igsf) {
-      ambigTracksInThisElectron.push_back(*igsf);
-    }
-
     // Now we need to clear them (they are the refs to original collection):
     gsfElectron.clearAmbiguousGsfTracks();
 
+    // Also in this loop let's relink ambiguous tracks
     // And here we add them back, now from a new reduced collection:
-    for (const auto& it : ambigTracksInThisElectron) {
+    for (const auto& it : gsfElectron.ambiguousGsfTracks()) {
```

Comment:
given the `clearAmbiguousGsfTracks` a few lines above, is the new version of the code indeed equivalent to the old?

Reply:
It seemed to me that the few deleted lines above just copied the `gsfElectron.ambiguousGsfTracks()` into a new vector `ambigTracksInThisElectron`, but we don't need to copy just for const access.

I agree, it seems suspicious because one wonders why the copying was done in the first place if it has no apparent purpose? I couldn't think of a reason.

Reply:
Absolutely not, you are right! I just see now the `gsfElectron.clearAmbiguousGsfTracks();` before. So the copying is needed. Will fix.

Reply:
I will fix the latest commit. Strange that it didn't show up in the matrix tests.